Scale Autonomous

Rohan-Sreenivasan opened the pull request:
This auto will put a cube on the scale, or switch or just cross the line depending on what readings we get.

strimper13 committed:
Changes to Auto and minute LED stuffs

"I always have other complaints" -Matt

strimper13 committed:
Changes to FIx Auto

Getting lift to move up based on driver station code.

strimper13 committed:
New Auto V.1

"That's Fine" -Matt 2K 18

strimper13 committed:
Formatted Auto

"It"s ready [Code} , but are they?" *Points to Mech with robot* -Matt

Rohan-Sreenivasan committed:
Scale Auto

This has autos that can do the scale, switch or just go forward, depending on what readings we get.

bot190 commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -7,8 +7,11 @@
 
 package frc.team166.robot;
 
+import javax.lang.model.util.ElementScanner6;
```
Lets get rid of this guy again

bot190 commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +135,90 @@ public void testPeriodic() {
+        gameData = DriverStation.getInstance().getGameSpecificMessage();
+        double degrees = 0.0;
+         if (gameData.length() > 0) {
+            if (gameData.charAt(1) == 'L') {
```
Should this be checking for 'R'?

bot190 commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +135,90 @@ public void testPeriodic() {
+                degrees = -90.00;
+                cmdRightAuto = new CommandChain("Mid Auto").then(drive.DriveTime(1.3, .6)).then(lift.GoToTopLift())
+                .then(drive.TurnByDegrees(degrees)).then(manipulator.CubeDrop());
+            } else  if (gameData.charAt(0) == 'L'){
```
Same here, this should be 'R'

bot190 commented on src/main/java/frc/team166/robot/subsystems/Lift.java:
```diff
@@ -323,6 +325,28 @@ protected boolean isFinished() {
+
+            @Override
+            protected boolean isFinished() {
+                engageBrake();
```
This function should be returning true or false to indicate the command is done. engageBrake() should go in end(), as well as a call to liftDrive.set(0) or something similar to ensure we stop the motor.

bot190 reviewed (commented):
This looks awesome guys! Just a few comments, but I really hope this works out.

Rohan-Sreenivasan committed:
Scale, switch and line cross auto fixed

Bunch of fixer uppers

bot190 commented on src/main/java/frc/team166/robot/subsystems/Lift.java:
```diff
@@ -206,6 +207,7 @@ protected boolean isFinished() {
             protected void end() {
                 liftDrive.stopMotor();
                 engageBrake();
+                liftDrive.set(0);
```
This is unnecessary, as the liftDrive.stopMotor() call above will do the same thing. The GoToTopLift command you added needs a end function like this one.

bot190 commented on src/main/java/frc/team166/robot/subsystems/Lift.java:
```diff
@@ -324,6 +327,31 @@ protected boolean isFinished() {
+
+            @Override
+            protected void execute() {
+                if (topLimitSwitch.get() == true) {
```
This if should not exist here. Instead the isFinished function below should be returning `!topLimitSwitch.get()` and there should be an end function that stops the drive motor and engages the brake

billcalabro replied:
Ben is right, and I think this will apply to the "GoToBottomLift" command as well... The Execute's for these should just be driving the lift up or down (positive or negative speeds). The IsFinished's should just be looking for either the top or bottom limit switch to be set. I can't remember if that digital input will return a true or a false when the switch is pressed, but you should be able to check that with SmartDash. Then in the IsFinished's all you need is `return topLimitSwitch.get()` or `return !topLimitSwitch.get()`

bot190 reviewed (changes requested):
The GoToTopLift command has some issues with the end logic.

Rohan-Sreenivasan committed:
Lift Fix

This command that brings the lift to the top needed some fixes.

Rohan-Sreenivasan committed:
UNH Fixes

Davin, Sean, Alex and I have been working hard fixing stuff. Mostly lift stuff.

Rohan-Sreenivasan committed:
fixes

more fixes

Rohan-Sreenivasan committed:
Fixed Auto????

We might have fixed the cross line and drop cube

bot190 commented on src/main/java/frc/team166/robot/subsystems/Lift.java:
```diff
@@ -215,6 +226,49 @@ protected void interrupted() {
+            protected boolean isFinished() {
+                return isTimedOut();
+            }
+        };
```
Don't forget to add an end() that calls engageBrake(); and liftDrive.stopMotor(). The command above should also have an end() function.

bot190 commented on src/main/java/frc/team166/robot/subsystems/Lift.java:
```diff
@@ -215,6 +226,49 @@ protected void interrupted() {
+
+            @Override
+            protected boolean isFinished() {
+                if (bottomLimitSwitch.get()) {
```
This can be `return bottomLimitSwitch.get();`, which is functionally equivalent

bot190 reviewed (changes requested):
Need to remember to add end() functions to stop the lift from moving.

Rohan-Sreenivasan committed:
We on that auto grind😤😤😤

just took three dubs on the field, bout to go four. MORE FIXES

Rohan-Sreenivasan committed:
Last fix for 3/30/18

End of the first day. still having trouble with the auto but WE WON 4 MATCHES IN A ROW!!!!!!!!

billcalabro commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+                cmdLeftAuto = new CommandChain("Cross Line Auto").then(drive.DriveTime(.3, .6));
+
+            }
+            return cmdLeftAuto;
```
Don't need this here, since you've got the same thing below. Once all the If-Else is done, it will return the cmdLeftAuto

billcalabro commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+        Command cmdLeftAuto = new CommandChain("Autonomous");
+        //The line above is to make VS Code happy and to take care of syntax errors
+        gameData = DriverStation.getInstance().getGameSpecificMessage();
+        double degrees = 0.0;
```
I would just get rid of this variable and pass "90.0" into your command chains directly

billcalabro commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+        if (gameData.length() > 0) {
+            if (gameData.charAt(1) == 'L') {
+                //we gon do the scale. Put in corner
+                degrees = 90.00;
```
remove this...

billcalabro commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+                //we gon do the scale. Put in corner
+                degrees = 90.00;
+                cmdLeftAuto = new CommandChain("Scale Auto").then(drive.DriveTime(1.3, .6)).then(lift.GoToTopLift())
+                        .then(drive.TurnByDegrees(degrees)).then(manipulator.CubeDrop());
```
...and here just do `.then(drive.TurnByDegrees(90.00)).then(manipulator.CubeDrop());`

billcalabro commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+                        .then(drive.TurnByDegrees(degrees)).then(manipulator.CubeDrop());
+            } else if (gameData.charAt(0) == 'L') {
+                //move forward and eject on switch. Align right behind switch
+                degrees = 90.00;
```
...same thing here, don't bother setting "degrees" just send that directly into your command chain

billcalabro commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+        return cmdLeftAuto;
+    }
+
+    public Command ScaleAutoRightStart() {
```
same comments from "ScaleAutoLeftStart" apply here...

billcalabro commented on src/main/java/frc/team166/robot/subsystems/Lift.java:
```diff
@@ -215,6 +227,50 @@ protected void interrupted() {
+            @Override
+            protected void initialize() {
+                disengageBrake();
+                setTimeout(.1);
```
Shouldn't this "0.1" be a parameter going into this function? That way we can change how long it runs every time we call it? Should be something like `public Command LowerLiftforTime(double duration)`

Rohan-Sreenivasan committed:
Matt just got here and we fixing stuff.

bot190 commented on src/main/java/frc/team166/robot/Robot.java:
```diff
@@ -132,8 +136,90 @@ public void testPeriodic() {
+
+        gameData = DriverStation.getInstance().getGameSpecificMessage();
+        double degrees = 0.0;
+        if (gameData.length() > 0) {
```
I just realized that gamedata will likely be empty when this function is called (At robot start time) As such the commands returned by this and the two scale functions will not be what is expected. The game data needs to be looked at during the execution of a command to ensure that the match has started and game data is available.

Rohan-Sreenivasan committed:
UNH FINAL

These are the final changes that I've made from UNH. We are yet to test the autos though.
